EventProc/src/OverlayProducer.cxx

```diff
@@ -0,0 +1,332 @@
+		  overlayHit.setTime( overlayTime ); 
+		  
+		  if ( isEcalHitCollection ) { //special treatment for ecal
+			int overlayHitID = overlayHit.getID();
+			if ( hitMap.find( overlayHitID ) == hitMap.end() ) { // there wasn't already a simhit in this id
+			  hitMap[overlayHitID] = SimCalorimeterHit();
+			  hitMap[ overlayHitID ].setID(overlayHitID);
+			  std::vector <float> hitPos = overlayHit.getPosition();
+			  hitMap[ overlayHitID ].setPosition( hitPos[0], hitPos[1], hitPos[2]);
+			}
+			// add the overlay hit (as a) contrib 
+			// incidentID = -1000, trackID = -1000, pdgCode = 0  <-- these are set in the header for now but could be parameters
+			hitMap[ overlayHitID ].addContrib(overlayIncidentID_,overlayTrackID_,overlayPdgCode_, overlayHit.getEdep(), overlayTime);
+		  } // ecal hits
+		  else {
+			std::map<std::string, int>::iterator itr = collectionMap.find( outCollName );
+			if (verbosity_ > 2)
+			  ldmx_log(debug) <<  "Adding non-Ecal overlay hit to outhit vector at " << itr->second ;
+			v_outHits[ itr->second ].push_back( overlayHit );
+		  }
+		} //over overlay calo simhit collection
+
+
+		if ( ! isEcalHitCollection )
+		  ldmx_log(debug) <<  "Nhits in overlay collection " << outCollName << ": " <<  v_outHits[ collectionMap.find( outCollName )->second ].size() ;
+		
+		
+	  } // over collections 
+	  
+	  if (! overlayFile_->nextEvent() ) {
+		ldmx_log(error) << "At sim event " << event.getEventHeader().getEventNumber() <<": couldn't read next overlay event!" ;
+		return;
+	  }
+	  
+	  
+	} // over overlay events 
+
+	
+	//after all events are done, the hitmap is final and can be written to the event output
+	for (uint iColl = 0; iColl < collections_.size(); iColl++) {
+	  // just loop through once to find the right collection name
+
+	  // add overlaid ecal hits as contribs/from hitmap rather than as copied simhits 
+	  if (strstr (collections_[iColl].c_str(), "Ecal") ) {
+
+		if (verbosity_ > 2)
+		  ldmx_log(debug) <<  "Hits in hitmap after overlay of " << collections_[iColl] << "Overlay :" ;
+		for ( auto &mapHit : hitMap ) {
+		  if (verbosity_ > 2)
+			mapHit.second.Print();
+
+		  if ( collectionMap.find(collections_[iColl]+"Overlay") == collectionMap.end() ) {
+			ldmx_log(debug) <<  "Adding first hit from hit map as first outhit vector to v_outHits";
+			v_outHits.push_back( {mapHit.second} );
+			collectionMap[ collections_[iColl]+"Overlay" ] = v_outHits.size() - 1;
+		  }
+		  else
+			v_outHits.back().push_back( mapHit.second );
+		}
+		
+	  }// isEcal
+	}//second loop over collections, to collect hits from hitmap
+
+	
+	// done collecting hits. 
+
+	// this should be added to the sim file, so to "event"
+	for (auto &itr : collectionMap ) {
+
+	  ldmx_log(debug) <<  "Writing " << itr.first << " to event bus" ;
+	  if (verbosity_ > 2) {
+		ldmx_log(debug) <<  "List of hits added: " ;
+		for ( auto &hit : v_outHits[ itr.second ] )
+		  hit.Print();
+	  }
+
+	  event.add( itr.first, v_outHits[ itr.second ]);
+	}
```

Comment:
A nice feature of C++17 which would clean up this code is python-like looping:
```
for (auto const& [ key , val ] : a_map ) {
    bool always_true_by_definition = (val == a_map.at(key));
}
```
If you were to combine `collectionMap` and `v_outHits` then this loop would look really nice:
```
for (auto const& [ name , coll ] : collectionMap ) {
    ldmx_log(debug) << "Writing " << name << " to event bus.";
    if (verbosity_ > 2) {
        ldmx_log(debug) << "List of hits added: ";
        for ( auto const& hit : coll ) hit.Print();
    }
    event.add( name , coll );
}
```